SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityPriorityKillPoller.java

```diff
@@ -0,0 +1,116 @@
+import com.google.inject.Inject;
+import com.google.inject.Singleton;
+import com.hubspot.mesos.JavaUtils;
+import com.hubspot.singularity.SingularityPendingTaskId;
+import com.hubspot.singularity.SingularityPriorityFreezeParent;
+import com.hubspot.singularity.SingularityRequestWithState;
+import com.hubspot.singularity.SingularityTaskCleanup;
+import com.hubspot.singularity.SingularityTaskId;
+import com.hubspot.singularity.TaskCleanupType;
+import com.hubspot.singularity.config.SingularityConfiguration;
+import com.hubspot.singularity.data.PriorityManager;
+import com.hubspot.singularity.data.RequestManager;
+import com.hubspot.singularity.data.TaskManager;
+
+@Singleton
+public class SingularityPriorityKillPoller extends SingularityLeaderOnlyPoller {
+    private static final Logger LOG = LoggerFactory.getLogger(SingularityScheduledJobPoller.class);
+
+    private final PriorityManager priorityManager;
+    private final RequestManager requestManager;
+    private final TaskManager taskManager;
+
+    private final double defaultTaskPriorityLevel;
+
+    @Inject
+    public SingularityPriorityKillPoller(PriorityManager priorityManager, RequestManager requestManager, TaskManager taskManager, SingularityConfiguration configuration) {
+        super(configuration.getCheckPriorityKillsEveryMillis(), TimeUnit.MILLISECONDS);
+
+        this.priorityManager = priorityManager;
+        this.requestManager = requestManager;
+        this.taskManager = taskManager;
+
+        this.defaultTaskPriorityLevel = configuration.getDefaultTaskPriorityLevel();
+    }
+
+
+    @Override
+    public void runActionOnPoll() {
+        if (!priorityManager.checkPriorityKillExists()) {
+            LOG.trace("No priority freeze to process.");
+            return;
+        }
+
+        final Optional<SingularityPriorityFreezeParent> maybePriorityFreeze = priorityManager.getActivePriorityFreeze();
+
+        if (!maybePriorityFreeze.isPresent() || !maybePriorityFreeze.get().getPriorityFreeze().isKillTasks()) {
+            LOG.trace("Priority freeze does not exist.");
+            priorityManager.clearPriorityKill();
+            return;
+        }
+
+        LOG.info("Handling priority freeze {}", maybePriorityFreeze.get());
+
+        final long now = System.currentTimeMillis();
+        int cancelledPendingTaskCount = 0;
+        int killedTaskCount = 0;
+
+        try {
+            final double minPriorityLevel = maybePriorityFreeze.get().getPriorityFreeze().getMinimumPriorityLevel();
+
+            // map request ID to priority level
+            final Map<String, Double> requestIdToTaskPriority = new HashMap<>();
+            for (SingularityRequestWithState requestWithState : requestManager.getRequests()) {
+                requestIdToTaskPriority.put(requestWithState.getRequest().getId(), requestWithState.getRequest().getTaskPriorityLevel().or(defaultTaskPriorityLevel));
+            }
+
+            // cancel pending tasks below minimum priority level
```

Comment:
I'd be worried about deleting pending tasks, because there won't necessarily be a pending request in the queue to put them back when we undo the freeze.  We would have to do something like we do on startup where we loop over all tasks. Maybe the scheduler should just be smart enough to ignore those pending tasks?